### PR TITLE
Notifications (architecture)

### PR DESCRIPTION
**Pull request by Patrick-Pataky**
This PR provides the basic architecture for real-time notifications support. 

It handles:
- Background notifications
- Foreground notifications

The Background notifications open the app in a specific page, while the Foreground notifications open a popup showing the event, along with a button that redirects to the proper page.

To send the notifications, the DB must contain a new field for every user: `fcmToken`. This PR already handles that. However, the current structure has a known limitation: 
- for each account, there can be only 1 device that can receive notifications at a time. 

The application asks for the Notification permissions right at the beginning of the application, and will soon be also available on the settings page. They are 3 main channels: Incoming/Outgoing/Social Events. The user can choose which notifications he/she wants to receive.

This PR provides a simple example of user flow notification (on foreground only):
- An user sends a new Loan Request (outgoing request)
- The owner of the item accepts/rejects the request (incoming request)

### CODE REVIEW

File: app/src/main/java/com/android/partagix/ui/App.kt [hunk 1]
```diff
@@ -162,20 +177,29 @@ class App(
...
-              Inventory(user.uid, emptyList()))
-      db.getUser(user.uid, { db.createUser(newUser) }, {})
-    }
-    // test that navigationActions has been initialized
+      notificationManager.checkToken(user.uid) { newToken ->
+        val newUser =
+            User(
+                user.uid,
+                user.displayName ?: "",
+                user.email ?: "",
```
Comment (LozanoCyrielle) [inline app/src/main/java/com/android/partagix/ui/App.kt:185]: Here, you need to leave the "Unknown Location field" or we will still have the Location as the email address``
```suggestion
                "Unknown Location",
```

File: app/src/main/java/com/android/partagix/model/notification/FirebaseMessagingService.kt [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+            Date.valueOf(data["creationDate"])
+          } catch (e: IllegalArgumentException) {
+            Date(System.currentTimeMillis())
+          }
+
+      val notification =
+          Notification(
+              title = notificationBody.title ?: "",
+              message = notificationBody.body ?: "",
+              type = Notification.Type.NEW_INCOMING_REQUEST,
```
Comment (LozanoCyrielle) [inline app/src/main/java/com/android/partagix/model/notification/FirebaseMessagingService.kt:105]: Here should it always be NEW_INCOMING_REQUEST

**Review by LozanoCyrielle — CHANGES_REQUESTED** [review]
The job you have done is insane. Great work !
Just some small changes :)) but again nice job !! 
The notification you have implemented in this PR are perfectly working :))

File: app/src/main/java/com/android/partagix/model/notification/FirebaseMessagingService.kt (hunk 2 above)
Comment (Patrick-Pataky) [reply app/src/main/java/com/android/partagix/model/notification/FirebaseMessagingService.kt]: You are right, the type should not be hardcoded, but collected from the notification content itself. This is fixed now

**Review by LozanoCyrielle — APPROVED** [review]
Looks perfect to me :))

**Review by skyrag — APPROVED** [review]
very good code, I have nothing to say. And the notifications work just fine, excellent job

**Comment by Patrick-Pataky** [comment]
Sorry for the modifications after the review. The code reviewed hasn't change, I only added new tests for coverage

**Pull request merged**